File: typescript/src/attribution/v3-convert.ts
import { createHash } from "node:crypto";

import {
  Decimal,
  canonicalDecimal,
  isoCanonical,
  type CostEvent,
} from "../core/models.js";
import {
  ATTRIBUTION_COMPONENTS,
  ATTRIBUTION_UNIT_BY_METRIC,
  type AttributionComponent,
  type AttributionUsageMetric,
} from "./types.js";
import {
  attributionComponentAndUsage,
  attributionEvidenceFor,
  attributionProviderFor,
  attributionResourceFor,
} from "./convert.js";
import type {
  AttributionBillingDimension,
  AttributionBillingDimensionValue,
  AttributionEventV3,
  AttributionOperationIdentityV3,
  AttributionOperationStatusV3,
  AttributionUsageLineV3,
} from "./v3-types.js";
import { validateAttributionObservationV3 } from "./v3-validate.js";

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const CANONICAL_NAME = /^[a-z0-9][a-z0-9._-]{0,127}$/;
const UNIT = /^[A-Za-z0-9][A-Za-z0-9._{}/*^+-]{0,63}$/;
const TRACE_ID = /^[0-9a-f]{32}$/;
const SPAN_ID = /^[0-9a-f]{16}$/;
const COMPONENTS = new Set<string>(ATTRIBUTION_COMPONENTS);

function stringDetail(details: Record<string, unknown>, ...keys: string[]): string | undefined {
  for (const key of keys) {
    const value = details[key];
    if (typeof value === "string" && value.trim() !== "") return value.trim();
  }
  return undefined;
}

function numberDetail(details: Record<string, unknown>, ...keys: string[]): number | undefined {
  for (const key of keys) {
    const value = details[key];
    if (typeof value === "number" && Number.isFinite(value)) return value;
    if (typeof value === "string" && value.trim() !== "") {
      const parsed = Number(value);
      if (Number.isFinite(parsed)) return parsed;
    }
  }
  return undefined;
}

function positiveQuantity(value: unknown): string | undefined {
  if (value === undefined || value === null) return undefined;
  try {
    const decimal = value instanceof Decimal ? value : new Decimal(String(value));
    if (!decimal.isFinite() || !decimal.gt(0)) return undefined;
    const rounded = decimal.toDecimalPlaces(12);
    return rounded.gt(0) ? canonicalDecimal(rounded) : undefined;
  } catch {
    return undefined;
  }
}

function deterministicUuid(namespace: string, ...parts: string[]): string {
  const bytes = createHash("sha256")
    .update([namespace, ...parts].join("\u0000"), "utf8")
    .digest()
    .subarray(0, 16);
  bytes[6] = (bytes[6] & 0x0f) | 0x50;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;
  const hex = bytes.toString("hex");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

function parseDimensionValue(value: unknown): AttributionBillingDimensionValue | undefined {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return undefined;
  const candidate = value as Record<string, unknown>;
  if (candidate.type === "string" && typeof candidate.value === "string" &&
      candidate.value.length > 0 && candidate.value.length <= 256) {
    return { type: "string", value: candidate.value };
  }
  if (candidate.type === "boolean" && typeof candidate.value === "boolean") {
    return { type: "boolean", value: candidate.value };
  }
  if (candidate.type === "integer" && typeof candidate.value === "string" &&
      /^-?(?:0|[1-9]\d{0,25})$/.test(candidate.value)) {
    return { type: "integer", value: candidate.value };
  }
  if (candidate.type === "decimal" && typeof candidate.value === "string" &&
      /^-?(?:0|[1-9]\d{0,25})(?:\.\d{1,12})?$/.test(candidate.value)) {
    return { type: "decimal", value: candidate.value };
  }
  return undefined;
}

function explicitDimensions(details: Record<string, unknown>): AttributionBillingDimension[] | null {
  const raw = details.attribution_dimensions;
  if (raw === undefined) return [];
  if (!Array.isArray(raw) || raw.length > 24) return null;
  const dimensions: AttributionBillingDimension[] = [];
  for (const candidate of raw) {
    if (candidate === null || typeof candidate !== "object" || Array.isArray(candidate)) return null;
    const dimension = candidate as Record<string, unknown>;
    if (typeof dimension.key !== "string" || !CANONICAL_NAME.test(dimension.key)) return null;
    const value = parseDimensionValue(dimension.value);
    if (value === undefined) return null;
    dimensions.push({ key: dimension.key, value });
  }
  dimensions.sort((left, right) => left.key.localeCompare(right.key));
  return dimensions;
}

function gpuSignalUsage(event: CostEvent): {
  component: AttributionComponent;
  usage: Array<{ metric: string; quantity: string; unit: string }>;
  durationSeconds?: number;
  dimensions: AttributionBillingDimension[];
} {
  const details = event.details;
  const candidates: Array<[string, unknown, string]> = [
    ["gpu.sm_utilization_percent", details.sm_util_pct, "Percent"],
    ["gpu.memory_utilization_percent", details.mem_util_pct, "Percent"],
    ["gpu.vram_peak_bytes", details.vram_used_peak_bytes, "Bytes"],
    ["gpu.vram_capacity_bytes", details.vram_total_bytes, "Bytes"],
    ["gpu.process_count", details.process_count, "Processes"],
    ["gpu.sample_count", details.sample_count, "Samples"],
  ];
  const usage = candidates.flatMap(([metric, rawQuantity, unit]) => {
    const quantity = positiveQuantity(rawQuantity);
    return quantity === undefined ? [] : [{ metric, quantity, unit }];
  });
  const dimensions: AttributionBillingDimension[] = [];
  const gpuIndex = numberDetail(details, "gpu_index");
  if (gpuIndex !== undefined && Number.isInteger(gpuIndex) && gpuIndex >= 0) {
    dimensions.push({ key: "gpu_index", value: { type: "integer", value: String(gpuIndex) } });
  }
  const sku = stringDetail(details, "gpu_sku");
  if (sku !== undefined) {
    dimensions.push({ key: "gpu_sku", value: { type: "string", value: sku.slice(0, 256) } });
  }
  const durationMs = numberDetail(details, "task_duration_ms");
  return {
    component: "gpu",
    usage,
    durationSeconds: durationMs !== undefined && durationMs > 0 ? durationMs / 1_000 : undefined,
    dimensions,
  };
}

function selectedComponent(
  event: CostEvent,
  fallback: AttributionComponent,
): AttributionComponent | null {
  const explicit = stringDetail(event.details, "attribution_component");
  if (explicit === undefined) return fallback;
  return COMPONENTS.has(explicit) ? explicit as AttributionComponent : null;
}

function operationName(event: CostEvent): string {
  const explicit = stringDetail(event.details, "attribution_operation_name");
  if (explicit !== undefined && CANONICAL_NAME.test(explicit)) return explicit;
  switch (event.eventType) {
    case "llm_call": return "llm.call";
    case "external_cost": return "external.call";
    case "compute_cost": return "compute.consume";
    case "gpu_cost": return "gpu.consume";
    case "gpu_utilization_signal": return "gpu.observe";
    case "network": return "network.transfer";
    case "retry_marker": return "retry.attempt";
  }
}

function operationStatus(event: CostEvent): AttributionOperationStatusV3 {
  const explicit = stringDetail(event.details, "attribution_operation_status");
  if (explicit === "in_progress" || explicit === "succeeded" || explicit === "failed" ||
      explicit === "cancelled" || explicit === "unknown") return explicit;
  if (event.eventType === "gpu_utilization_signal") return "unknown";
  if (event.eventType === "retry_marker" || stringDetail(event.details, "error_type") !== undefined) {
    return "failed";
  }
  return "succeeded";
}

function operationFor(event: CostEvent): AttributionOperationIdentityV3 | null {
  const explicitOperationId = stringDetail(event.details, "attribution_operation_id");
  if (event.retryOf !== undefined &&
      (typeof event.retryOf !== "string" || !UUID.test(event.retryOf))) return null;
  const retryOf = event.retryOf?.toLowerCase();
  const hasValidOperationId = explicitOperationId !== undefined && UUID.test(explicitOperationId);
  const operationId = hasValidOperationId
    ? explicitOperationId!.toLowerCase()
    : event.eventId.toLowerCase();
  const explicitAttemptId = stringDetail(event.details, "attribution_attempt_id");
  const attemptId = explicitAttemptId !== undefined && UUID.test(explicitAttemptId)
    ? explicitAttemptId.toLowerCase()
    : event.eventId.toLowerCase();
  const explicitAttempt = numberDetail(event.details, "attribution_attempt_number");
  const hasValidAttemptNumber = explicitAttempt !== undefined && Number.isInteger(explicitAttempt) &&
    explicitAttempt > 0;
  if (retryOf !== undefined &&
      (!hasValidOperationId || !hasValidAttemptNumber || (explicitAttempt ?? 0) <= 1)) return null;
  const attemptNumber = hasValidAttemptNumber ? explicitAttempt! : 1;
  const operation: AttributionOperationIdentityV3 = {
    id: operationId,
    name: operationName(event),
    status: operationStatus(event),
    attempt: { id: attemptId, number: attemptNumber },
  };
  if (retryOf !== undefined) operation.attempt.retry_of = retryOf;
  const traceId = stringDetail(event.details, "trace_id")?.toLowerCase();
  const spanId = stringDetail(event.details, "span_id")?.toLowerCase();
  if (traceId !== undefined && spanId !== undefined && TRACE_ID.test(traceId) && SPAN_ID.test(spanId)) {
    operation.trace = { trace_id: traceId, span_id: spanId };
  }
  return operation;
}

function unknownExplicitUsage(event: CostEvent): {
  component: AttributionComponent;
  usage: Array<{ metric: string; quantity: string; unit: string }>;
  durationSeconds?: number;
} | undefined {
  if (event.eventType !== "external_cost") return undefined;
  const metric = stringDetail(event.details, "attribution_usage_metric");
  if (metric === undefined || metric in ATTRIBUTION_UNIT_BY_METRIC || !CANONICAL_NAME.test(metric)) {
    return undefined;
  }
  const unit = stringDetail(event.details, "attribution_usage_unit");
  const quantity = positiveQuantity(event.details.attribution_usage_quantity);
  if (unit === undefined || !UNIT.test(unit) || quantity === undefined) return undefined;
  return {
    component: "external",
    usage: [{ metric, quantity, unit }],
    durationSeconds: numberDetail(event.details, "attribution_usage_duration_seconds"),
  };
}

/** Convert durable v1 capture into the strict, details-free v3 observation. */
export function toAttributionObservationV3(event: CostEvent): AttributionEventV3 | null {
  const explicit = unknownExplicitUsage(event);
  const gpuSignal = event.eventType === "gpu_utilization_signal" ? gpuSignalUsage(event) : undefined;
  const legacy = explicit === undefined && gpuSignal === undefined
    ? attributionComponentAndUsage(event)
    : undefined;
  const mapped = explicit ?? gpuSignal ?? legacy;
  if (mapped === null || mapped === undefined) return null;

  const explicitBillingDimensions = explicitDimensions(event.details);
  if (explicitBillingDimensions === null) {
    console.warn(`[dexcost] Event ${event.eventId} has invalid attribution_dimensions`);
    return null;
  }
  const dimensions = gpuSignal === undefined
    ? explicitBillingDimensions
    : [...explicitBillingDimensions, ...gpuSignal.dimensions]
      .sort((left, right) => left.key.localeCompare(right.key));
  const stableDimensions = JSON.stringify(dimensions);
  const usage: AttributionUsageLineV3[] = mapped.usage.map((line) => ({
    line_id: deterministicUuid(
      "dexcost:attribution-usage-line:v3",
      event.eventId.toLowerCase(),
      line.metric,
      line.unit,
      stableDimensions,
    ),
    metric: line.metric,
    quantity: line.quantity,
    unit: line.unit,
    dimensions,
  }));
  const component = selectedComponent(event, mapped.component);
  if (component === null) {
    console.warn(`[dexcost] Event ${event.eventId} has an invalid attribution_component`);
    return null;
  }
  const occurredAt = isoCanonical(event.occurredAt);
  const operation = operationFor(event);
  if (operation === null) {
    console.warn(`[dexcost] Event ${event.eventId} has invalid or incomplete retry lineage`);
    return null;
  }
  const converted: AttributionEventV3 = {
    schema_version: "3",
    event_id: event.eventId.toLowerCase(),
    task_id: event.taskId.toLowerCase(),
    occurred_at: occurredAt,
    observed_at: occurredAt,
    component,
    provider: attributionProviderFor(event),
    operation,
    lifecycle: { state: "final", revision: 1 },
    usage_snapshot: "full",
    usage,
  };
  const resource = attributionResourceFor(event);
  if (resource !== undefined) converted.resource = resource;
  if (event.eventType !== "gpu_utilization_signal") {
    const evidence = attributionEvidenceFor(event);
    if (evidence !== undefined) converted.cost_evidence = evidence;
  }
  const hasTimeBasedUsage = usage.some((line) =>
    line.metric in ATTRIBUTION_UNIT_BY_METRIC &&
    line.unit === ATTRIBUTION_UNIT_BY_METRIC[line.metric as AttributionUsageMetric] &&
    line.unit.endsWith("Seconds"),
  );
  if (hasTimeBasedUsage || (mapped.durationSeconds !== undefined && mapped.durationSeconds > 0)) {
    const durationMs = mapped.durationSeconds !== undefined && mapped.durationSeconds > 0
      ? mapped.durationSeconds * 1_000
      : 0;
    converted.usage_period = {
      start_at: isoCanonical(new Date(event.occurredAt.getTime() - durationMs)),
      end_at: occurredAt,
    };
  }

  const validation = validateAttributionObservationV3(converted);
  if (!validation.success) {
    console.warn(
      `[dexcost] Event ${event.eventId} cannot be represented by attribution v3: ` +
      validation.issues.map((issue) => issue.path || "<root>").join(", "),
    );
    return null;
  }
  return converted;
}

export const toAttributionEventV3 = toAttributionObservationV3;

File: go/attribution/v3_convert.go
package attribution

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/DexwoxBusiness/dexcost-sdk/go/core"
	"github.com/shopspring/decimal"
)

var (
	v3UUIDPattern = regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)
	v3UnitPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._{}/*^+\-]{0,63}$`)
	v3TraceID     = regexp.MustCompile(`^[0-9a-f]{32}$`)
	v3SpanID      = regexp.MustCompile(`^[0-9a-f]{16}$`)
	v3Integer     = regexp.MustCompile(`^-?(?:0|[1-9]\d{0,25})$`)
	v3Decimal     = regexp.MustCompile(`^-?(?:0|[1-9]\d{0,25})(?:\.\d{1,12})?$`)
)

var v3Components = map[string]Component{
	string(ComponentLLM):               ComponentLLM,
	string(ComponentTelephony):         ComponentTelephony,
	string(ComponentVoicePlatform):     ComponentVoicePlatform,
	string(ComponentSpeechToText):      ComponentSpeechToText,
	string(ComponentTextToSpeech):      ComponentTextToSpeech,
	string(ComponentRealtimeTransport): ComponentRealtimeTransport,
	string(ComponentRecording):         ComponentRecording,
	string(ComponentPostCallAnalysis):  ComponentPostCallAnalysis,
	string(ComponentCompute):           ComponentCompute,
	string(ComponentGPU):               ComponentGPU,
	string(ComponentNetwork):           ComponentNetwork,
	string(ComponentStorage):           ComponentStorage,
	string(ComponentExternal):          ComponentExternal,
}

type v3MappedUsage struct {
	component       Component
	usage           []UsageLineV2
	durationSeconds decimal.Decimal
	dimensions      []BillingDimensionV3
}

func canonicalV3DimensionsJSON(dimensions []BillingDimensionV3) ([]byte, error) {
	var buffer bytes.Buffer
	encoder := json.NewEncoder(&buffer)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(dimensions); err != nil {
		return nil, err
	}
	encoded := bytes.TrimSuffix(buffer.Bytes(), []byte{'\n'})
	encoded = bytes.ReplaceAll(encoded, []byte(`\u2028`), []byte("\u2028"))
	encoded = bytes.ReplaceAll(encoded, []byte(`\u2029`), []byte("\u2029"))
	return encoded, nil
}

func deterministicV3UUID(namespace string, parts ...string) string {
	hash := sha256.New()
	values := append([]string{namespace}, parts...)
	for index, value := range values {
		if index > 0 {
			hash.Write([]byte{0})
		}
		hash.Write([]byte(value))
	}
	bytes := hash.Sum(nil)[:16]
	bytes[6] = (bytes[6] & 0x0f) | 0x50
	bytes[8] = (bytes[8] & 0x3f) | 0x80
	hexValue := hex.EncodeToString(bytes)
	return fmt.Sprintf("%s-%s-%s-%s-%s", hexValue[:8], hexValue[8:12], hexValue[12:16], hexValue[16:20], hexValue[20:])
}

func v3PositiveQuantity(value interface{}) (string, bool) {
	details := map[string]interface{}{"value": value}
	parsed, ok := decimalDetail(details, "value")
	if !ok {
		return "", false
	}
	return positiveQuantity(parsed)
}

func parseV3DimensionValue(value interface{}) (BillingDimensionValueV3, bool) {
	candidate, ok := value.(map[string]interface{})
	if !ok {
		return BillingDimensionValueV3{}, false
	}
	kind, _ := candidate["type"].(string)
	switch kind {
	case "string":
		text, ok := candidate["value"].(string)
		return BillingDimensionValueV3{Type: kind, Value: text}, ok && len(text) > 0 && len(text) <= 256
	case "boolean":
		boolean, ok := candidate["value"].(bool)
		return BillingDimensionValueV3{Type: kind, Value: boolean}, ok
	case "integer", "decimal":
		text, ok := candidate["value"].(string)
		valid := v3Integer.MatchString(text)
		if kind == "decimal" {
			valid = v3Decimal.MatchString(text)
		}
		return BillingDimensionValueV3{Type: kind, Value: text}, ok && valid
	default:
		return BillingDimensionValueV3{}, false
	}
}

func explicitV3Dimensions(details map[string]interface{}) ([]BillingDimensionV3, bool) {
	raw, exists := details["attribution_dimensions"]
	if !exists {
		return []BillingDimensionV3{}, true
	}
	encoded, err := json.Marshal(raw)
	if err != nil {
		return nil, false
	}
	var candidates []map[string]interface{}
	decoder := json.NewDecoder(strings.NewReader(string(encoded)))
	decoder.UseNumber()
	if err := decoder.Decode(&candidates); err != nil || len(candidates) > 24 {
		return nil, false
	}
	dimensions := make([]BillingDimensionV3, 0, len(candidates))
	for _, candidate := range candidates {
		key, ok := candidate["key"].(string)
		if !ok || !canonicalPattern.MatchString(key) {
			return nil, false
		}
		value, ok := parseV3DimensionValue(candidate["value"])
		if !ok {
			return nil, false
		}
		dimensions = append(dimensions, BillingDimensionV3{Key: key, Value: value})
	}
	sort.Slice(dimensions, func(i, j int) bool { return dimensions[i].Key < dimensions[j].Key })
	return dimensions, true
}

func gpuSignalV3Usage(event core.Event) v3MappedUsage {
	type candidate struct{ metric, key, unit string }
	candidates := []candidate{
		{"gpu.sm_utilization_percent", "sm_util_pct", "Percent"},
		{"gpu.memory_utilization_percent", "mem_util_pct", "Percent"},
		{"gpu.vram_peak_bytes", "vram_used_peak_bytes", "Bytes"},
		{"gpu.vram_capacity_bytes", "vram_total_bytes", "Bytes"},
		{"gpu.process_count", "process_count", "Processes"},
		{"gpu.sample_count", "sample_count", "Samples"},
	}
	usage := make([]UsageLineV2, 0, len(candidates))
	for _, item := range candidates {
		quantity, ok := v3PositiveQuantity(event.Details[item.key])
		if ok {
			usage = append(usage, UsageLineV2{Metric: UsageMetric(item.metric), Quantity: quantity, Unit: UsageUnit(item.unit)})
		}
	}
	dimensions := make([]BillingDimensionV3, 0, 2)
	if gpuIndex, ok := decimalDetail(event.Details, "gpu_index"); ok && !gpuIndex.IsNegative() && gpuIndex.Equal(gpuIndex.Truncate(0)) {
		dimensions = append(dimensions, BillingDimensionV3{Key: "gpu_index", Value: BillingDimensionValueV3{Type: "integer", Value: gpuIndex.String()}})
	}
	if sku := strings.TrimSpace(stringDetail(event.Details, "gpu_sku")); sku != "" {
		dimensions = append(dimensions, BillingDimensionV3{Key: "gpu_sku", Value: BillingDimensionValueV3{Type: "string", Value: truncate(sku, 256)}})
	}
	duration := decimal.Zero
	if durationMS, ok := decimalDetail(event.Details, "task_duration_ms"); ok && durationMS.IsPositive() {
		duration = durationMS.Div(decimal.NewFromInt(1_000))
	}
	return v3MappedUsage{component: ComponentGPU, usage: usage, durationSeconds: duration, dimensions: dimensions}
}

func unknownExplicitV3Usage(event core.Event) (v3MappedUsage, bool) {
	if event.EventType != core.EventTypeExternalCost {
		return v3MappedUsage{}, false
	}
	metric := strings.TrimSpace(stringDetail(event.Details, "attribution_usage_metric"))
	if metric == "" || !canonicalPattern.MatchString(metric) {
		return v3MappedUsage{}, false
	}
	if _, known := UnitByMetric[UsageMetric(metric)]; known {
		return v3MappedUsage{}, false
	}
	unit := strings.TrimSpace(stringDetail(event.Details, "attribution_usage_unit"))
	quantity, ok := v3PositiveQuantity(event.Details["attribution_usage_quantity"])
	if !ok || !v3UnitPattern.MatchString(unit) {
		return v3MappedUsage{}, false
	}
	duration, _ := decimalDetail(event.Details, "attribution_usage_duration_seconds")
	return v3MappedUsage{
		component:       ComponentExternal,
		usage:           []UsageLineV2{{Metric: UsageMetric(metric), Quantity: quantity, Unit: UsageUnit(unit)}},
		durationSeconds: duration,
	}, true
}

func selectedV3Component(event core.Event, fallback Component) (Component, bool) {
	explicit := strings.TrimSpace(stringDetail(event.Details, "attribution_component"))
	if explicit == "" {
		return fallback, true
	}
	component, ok := v3Components[explicit]
	return component, ok
}

func v3OperationName(event core.Event) string {
	explicit := strings.TrimSpace(stringDetail(event.Details, "attribution_operation_name"))
	if canonicalPattern.MatchString(explicit) {
		return explicit
	}
	switch event.EventType {
	case core.EventTypeLLMCall:
		return "llm.call"
	case core.EventTypeExternalCost:
		return "external.call"
	case core.EventTypeComputeCost:
		return "compute.consume"
	case core.EventTypeGPUCost:
		return "gpu.consume"
	case core.EventTypeGPUUtilizationSignal:
		return "gpu.observe"
	case core.EventTypeNetwork:
		return "network.transfer"
	case core.EventTypeRetryMarker:
		return "retry.attempt"
	default:
		return "external.call"
	}
}

func v3OperationStatus(event core.Event) string {
	explicit := strings.TrimSpace(stringDetail(event.Details, "attribution_operation_status"))
	switch explicit {
	case "in_progress", "succeeded", "failed", "cancelled", "unknown":
		return explicit
	}
	if event.EventType == core.EventTypeGPUUtilizationSignal {
		return "unknown"
	}
	if event.EventType == core.EventTypeRetryMarker || event.ErrorType != "" || stringDetail(event.Details, "error_type") != "" {
		return "failed"
	}
	return "succeeded"
}

func v3OperationFor(event core.Event) (OperationIdentityV3, bool) {
	eventID := strings.ToLower(event.EventID.String())
	retryOf := ""
	if event.RetryOf != nil {
		retryOf = strings.ToLower(event.RetryOf.String())
		if !v3UUIDPattern.MatchString(retryOf) {
			return OperationIdentityV3{}, false
		}
	}
	explicitOperationID := strings.TrimSpace(stringDetail(event.Details, "attribution_operation_id"))
	hasOperationID := v3UUIDPattern.MatchString(explicitOperationID)
	operationID := eventID
	if hasOperationID {
		operationID = strings.ToLower(explicitOperationID)
	}
	attemptID := eventID
	if explicit := strings.TrimSpace(stringDetail(event.Details, "attribution_attempt_id")); v3UUIDPattern.MatchString(explicit) {
		attemptID = strings.ToLower(explicit)
	}
	attemptNumber := int64(1)
	hasAttemptNumber := false
	if raw, ok := decimalDetail(event.Details, "attribution_attempt_number"); ok && raw.IsPositive() && raw.Equal(raw.Truncate(0)) && raw.LessThanOrEqual(decimal.NewFromInt(math.MaxInt32)) {
		attemptNumber = raw.IntPart()
		hasAttemptNumber = true
	}
	if retryOf != "" && (!hasOperationID || !hasAttemptNumber || attemptNumber <= 1) {
		return OperationIdentityV3{}, false
	}
	operation := OperationIdentityV3{
		ID:      operationID,
		Name:    v3OperationName(event),
		Status:  v3OperationStatus(event),
		Attempt: AttemptIdentityV3{ID: attemptID, Number: int(attemptNumber), RetryOf: retryOf},
	}
	traceID := strings.ToLower(strings.TrimSpace(stringDetail(event.Details, "trace_id")))
	spanID := strings.ToLower(strings.TrimSpace(stringDetail(event.Details, "span_id")))
	if v3TraceID.MatchString(traceID) && v3SpanID.MatchString(spanID) {
		operation.Trace = &TraceIdentityV3{TraceID: traceID, SpanID: spanID}
	}
	return operation, true
}

// ToObservationV3 converts durable v1 capture into strict attribution v3.
// Invalid or incomplete records return nil and are quarantined by transport.
func ToObservationV3(event core.Event) *ObservationV3 {
	mapped, explicit := unknownExplicitV3Usage(event)
	if event.EventType == core.EventTypeGPUUtilizationSignal {
		mapped = gpuSignalV3Usage(event)
		explicit = true
	}
	if !explicit {
		component, usage, duration, ok := componentAndUsage(event)
		if !ok {
			return nil
		}
		mapped = v3MappedUsage{component: component, usage: usage, durationSeconds: duration}
	}

	dimensions, ok := explicitV3Dimensions(event.Details)
	if !ok {
		log.Printf("[dexcost] event %s has invalid attribution_dimensions", event.EventID)
		return nil
	}
	dimensions = append(dimensions, mapped.dimensions...)
	sort.Slice(dimensions, func(i, j int) bool { return dimensions[i].Key < dimensions[j].Key })
	stableDimensions, err := canonicalV3DimensionsJSON(dimensions)
	if err != nil {
		return nil
	}
	usage := make([]UsageLineV3, 0, len(mapped.usage))
	for _, line := range mapped.usage {
		usage = append(usage, UsageLineV3{
			LineID: deterministicV3UUID("dexcost:attribution-usage-line:v3", strings.ToLower(event.EventID.String()), string(line.Metric), string(line.Unit), string(stableDimensions)),
			Metric: string(line.Metric), Quantity: line.Quantity, Unit: string(line.Unit), Dimensions: dimensions,
		})
	}
	component, ok := selectedV3Component(event, mapped.component)
	if !ok {
		log.Printf("[dexcost] event %s has invalid attribution_component", event.EventID)
		return nil
	}
	operation, ok := v3OperationFor(event)
	if !ok {
		log.Printf("[dexcost] event %s has invalid or incomplete retry lineage", event.EventID)
		return nil
	}
	occurred := canonicalTime(event.OccurredAt)
	converted := &ObservationV3{
		SchemaVersion: "3", EventID: strings.ToLower(event.EventID.String()), TaskID: strings.ToLower(event.TaskID.String()),
		OccurredAt: occurred, ObservedAt: occurred, Component: component, Provider: providerFor(event), Resource: resourceFor(event),
		Operation: operation, Lifecycle: LifecycleV3{State: "final", Revision: 1}, UsageSnapshot: "full", Usage: usage,
	}
	if event.EventType != core.EventTypeGPUUtilizationSignal {
		converted.CostEvidence = evidenceFor(event)
	}
	hasTime := false
	for _, line := range usage {
		canonical, known := UnitByMetric[UsageMetric(line.Metric)]
		if known && string(canonical) == line.Unit && strings.HasSuffix(line.Unit, "Seconds") {
			hasTime = true
			break
		}
	}
	if hasTime || mapped.durationSeconds.IsPositive() {
		offset := time.Duration(0)
		if mapped.durationSeconds.IsPositive() {
			micros := mapped.durationSeconds.Mul(decimal.NewFromInt(1_000_000)).Round(0).IntPart()
			offset = time.Duration(micros) * time.Microsecond
		}
		converted.UsagePeriod = &UsagePeriodV3{StartAt: canonicalTime(event.OccurredAt.Add(-offset)), EndAt: occurred}
	}
	validation := ValidateObservationV3(converted)
	if !validation.Success {
		paths := make([]string, 0, len(validation.Issues))
		for _, issue := range validation.Issues {
			paths = append(paths, issue.Path)
		}
		log.Printf("[dexcost] event %s cannot be represented by attribution v3: %s", event.EventID, strings.Join(paths, ", "))
		return nil
	}
	return converted
}

func ToEventV3(event core.Event) *EventV3 { return ToObservationV3(event) }
